### DISCUSSION TITLE
데이터베이스 정규화 과정1

### DISCUSSION THREAD

**Discussion by sujin-park0607** [24tngus/CS_STUDY#35] [Database]
# 정규화 과정

릴레이션 간의 잘못된 함수 종속 관계로 인해 데이터베이스 **이상 현상이 일어나는 것을 해결하는 과정**이자 **공간을 효율적으로 사용**하기 위해 릴레이션을 여러 개로 분리하는 과정

##  이상현상(anomaly)
데이터베이스의 설계를 잘못되었을 때 데이터 중복 발생, 삽입, 갱신, 삭제 연산을 수행할 때 부작용이 발생하는 것 
   
  > **삽입이상**:불필요한 데이터도 같이 삽입되는 문제
  > 
  > **갱신이상**:중복 레코드가 있고 이 중 일부만 변경해 데이터 자체가 불일 치 되는 문제
  > 
  > **삭제이상**:레코드를 삭제하는데 필요한 데이터까지 함께 삭제되는 문제
  >
  >  모두 함수의 종속 관계로 인해 발생한 문제!!  

    💡 함수의 종속 관계란?
    정규화를 통해 테이블을 분할할 때 어떻게 분할할 것인가? 그 기준이 무엇인가?
    이상현상을 해결하는척도가 되는 것이 **함수 종속성**
    

## 함수의 종속성
x 가 y를 결정한다.→ x가 결정자,  Y종속자
하나의 x에 대해  y가 하나라면 함수를 결정한다 종속되어 있다는 관계를 가진다.
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/e6b22416-c86f-480a-aa3b-df191b0552b3)
화살표 나가는 곳 : 결정자
화살표 도착하는 곳: 종속자 

## 완전함수종속 부분함수종속
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/d6c50d6f-27aa-489c-b97d-427adbd90d6c)
{아이디, 이벤트번호} → 당첨여부
아이디 → 이름

![image](https://github.com/24tngus/CS_STUDY/assets/75667075/bfbe1273-446d-43e0-8449-8ad951443e85)
완전함수 종속: Y가 X전체에 종속된 관계 ex) 당첨여부
부분함수 종속: Y가 X일부에 종속된 것 ex) 이름

### 제1정규형(1FN)

릴레이션의 속성 값 중에서 한 개의 기본키에 대해 두 개 이상의 값을 가지는 반복 집합이 있다면 제거하는 것을 말합니다.
**한개의 → 하나의 원자 값** 
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/c2c532af-bd73-49c4-8b25-b5bd59d8fa72)

### 제2정규형(2FN)
기본키가 아닌 모든 속성이 기본키에 완전함수종속으로 만들어놓는것이다.
**before**
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/0b55d9c2-dafd-473c-a568-9d21621899a3)
{아이디, 이벤트번호} → 당첨여부
아이디 → 등급 → 할인율

**after**
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/d008ec97-3739-4dfc-a470-06421cc6c33e)
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/b541132d-dac8-44b2-850e-25f51c591ca1)
- 분해한 릴레이션은 조인이 가능해야하고 조인되었을 때 분해 전의 릴레이션들로 복원이 가능해야함.
- 즉, 분해했을때 정보 손실이 없는 **무손실분해**가 되어야함

### 제3정규형(3FN)
제 2 정규형이고 기본키가 아닌 모든 속성이 **기본키에 이행적 함수 종속을** 만족하지 않는 상태

이행적 함수 종속
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/122d625c-b7c1-412d-a177-fe06e7b7315f)

**before** : 아이디 -> 등급 -> 할인율
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/75ca3df0-38d8-46ca-b006-ac62f2cb576b)

**after** : 
- 아이디 -> 등급
- 등급 -> 할인율
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/ae23190c-68d5-44e1-9a2e-a081e68effe0)

### 보이스/코드 정규형(BCNF)

모든 결정자가 후보키인 상태
보통 제3정규형보다 엄격한 기준으로 제약을 건 것

> 요구사항
> 각 수강명에 대해 한 학생은 오직 한 강사의 강의만 수강한다.
> 각 강사는 한 수강명만 담당한다
> 한 수강명은 여러 강사가 담당할 수 있다.

![image](https://github.com/24tngus/CS_STUDY/assets/75667075/f3960616-4a2c-45bf-ab1b-a13e9c6322f4)
강사는 후보키가 아님에도 불구하고 수강명을 결정하는 결정자 역할을 하게 된다.

**before**
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/ee7089dc-0034-4129-8500-0c6e84eee24e)
> 만약 범석이라는 강사가 '롤'이라는 수강명을 담당했을 때, 현재 테이블에 삽입하려 하면 학번이
> NULL이 되는 문제점이 발생한다. 

**after**
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/4347679c-1aa9-4d38-84b4-354b349086d4)

### 정규화과정 정리
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/c9ead62c-a2fc-4a1f-906f-ab3fd471149c)

비정규형 → 제1정규형 → 제2정규형 → 제3정규형 → 보이스/코드 정규형 

> 외우는 TIP! 
> **도부이결다조!!!!**

### 정규화 장점
- 데이터 구조의 안정성 유지
- 중복 값 & 널 값이 줄어든다
- 이상현상 제거
- 저장 공간 최소화

정규화 과정은 중복 데이터가 제거됨에 따라 스토리지에 대한 메모리 요구량이 감소하게 된다. 또한 테이블이 더 정상적이게 되므로 오류가 발생할 가능성이 낮다

> 💡 정규형과정이 꼭 필요한가?
> 어느정도는 필요하다! 정규형과정은 테이블릴레이션을 나누는 과정 
> 비즈니스 로직에 따라 조인이 많이 들어가게 되면 오히려 성능이 낮아질 수 있다
> 그래서 **반정규화과정**도 존재

**Comment by jindream6128** [comment]
이상현상 면접 필수 질문인거 같아요... ㅠ,ㅠ

**Comment by 24tngus** [comment]
확인했습니다

**Comment by berryberries** [comment]
확인했습니다!
